drivers: spi: andes_atcspi200: Support dma mode

kevinwang821020 opened the pull request:
1. Support the dma mode for andes_atcspi200 and use board adp_xc7k_ae350 for testing.
2. Refine the function mechanism.

Commit:
drivers: spi: Refine some coding style for andes_atcspi200

1. Remove the redundant code.
2. Use sys_set_bits and sys_clear_bits instead of customized MACRO.

Signed-off-by: Kevin Wang <kevinwang821020@google.com>

kevinwang821020 commented:
Sorry to bother you reviewers, I would like to know if there are any  parts of source code that I can improve/refined.

Thanks for helping us.

teburd commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
+			&data->dma_rx.dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		return ret;
```
Should this free the allocations on failure here?

teburd commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
+	spi_rx_dma_disable(spi_dev);
+
+	if (spi_context_rx_on(ctx)) {
+		spi_dma_rx_load(spi_dev);
```
the error return isn't caught here, (void)spi_dma_rx_load(spi_dev) to signal the return is ignored perhaps

teburd commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
+	if (spi_context_rx_on(ctx)) {
+		spi_dma_rx_load(spi_dev);
+		spi_rx_dma_enable(spi_dev);
+		dma_start(data->dma_rx.dma_dev, data->dma_rx.channel);
```
dma_start is fallible as well

teburd commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
+	spi_tx_dma_disable(spi_dev);
+
+	if (spi_context_tx_on(ctx)) {
+		spi_dma_tx_load(spi_dev);
```
fallible and the error is ignored, probably needs at least an assert or a (void) prefix

kevinwang821020 replied:
Thanks for your suggestion. 
We have add some function failed detection for dma_start.

teburd reviewed (changes requested):
Please make sure you run this against the spi test suite in some manner, there are many edge cases with the spi_buf_set that can be given to this API, would be nice to see a post here showing the test passing with this driver on hardware. It's not a requirement from me, but would give a good starting point showing the driver worked as of this point in time.

ycsin commented:
@kevinwang821020 any update on this one?

kevinwang821020 commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -12,14 +12,39 @@
+#define ANDES_SPI_DMA_DONE_FLAG	\
+	(ANDES_SPI_DMA_RX_DONE_FLAG | ANDES_SPI_DMA_TX_DONE_FLAG)
+
+K_HEAP_DEFINE(chain_block, 2048);
```
Sorry for replying so late. Due to the modification for the request, there are some bug issue, so we took a while to fixed then.
We have update the source code, thanks for your suggestion.

kevinwang821020 commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,30 +190,386 @@ static int configure(const struct device *dev,
+			&data->dma_tx.dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		return ret;
```
Thank your suggestion, we did not notice that.
After the modification for previous request, this issue has been improved.

ycsin commented on drivers/spi/Kconfig.andes_atcspi200:
```diff
@@ -11,3 +11,12 @@ config SPI_ANDES_ATCSPI200
+
+if SPI_ANDES_ATCSPI200
+
+	config ANDES_SPI_DMA_MODE
```
```suggestion
config ANDES_SPI_DMA_MODE
```

ycsin commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
+ * dummy value used for transferring NOP when tx buf is null
+ * and use as dummy sink for when rx buf is null
+ */
+uint32_t dummy_rx_tx_buffer;
```
will this work for multi instances? can we just move this into device data / runtime stack?

kevinwang821020 replied:
Yes, it work for multi instances, the data send/to this variable is noused.
I think move into runtime stack may cause some problem, because after the spi_dma_tx/rx_load function has return, the variable will be release, after that the address may be used by other function. But at that time, the transfer may still ongoing, so the variable used by other function in the address will be polluted.

ycsin commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -195,23 +607,58 @@ static int transceive(const struct device *dev,
+		data->chunk_len = chunk_len;
 
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+		if ((data->dma_tx.dma_dev != NULL)
+		 && (data->dma_rx.dma_dev != NULL)) {
```
the new column limit is 100, this should fit:
```suggestion
		if ((data->dma_tx.dma_dev != NULL) && (data->dma_rx.dma_dev != NULL)) {
```

ycsin commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -368,28 +829,104 @@ static void spi_atcspi200_irq_handler(void *arg)
+		sys_write32(0, SPI_INTEN(cfg->base));
+
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+		if ((data->dma_tx.dma_dev != NULL)
+		 && data->dma_tx.dma_cfg.source_chaining_en) {
```
```suggestion
		if ((data->dma_tx.dma_dev != NULL) && data->dma_tx.dma_cfg.source_chaining_en) {
```

ycsin commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -368,28 +829,104 @@ static void spi_atcspi200_irq_handler(void *arg)
+			data->dma_tx.dma_blk_cfg.next_block = NULL;
+		}
+
+		if ((data->dma_rx.dma_dev != NULL)
+		 && data->dma_rx.dma_cfg.source_chaining_en) {
```
```suggestion
		if ((data->dma_rx.dma_dev != NULL) && data->dma_rx.dma_cfg.source_chaining_en) {
```

kevinwang821020 replied:
Thanks for your suggestion!
We have update the source code.

nashif commented:
@kevinwang821020 can you please review?

kevinwang821020 commented:
> Please make sure you run this against the spi test suite in some manner, there are many edge cases with the spi_buf_set that can be given to this API, would be nice to see a post here showing the test passing with this driver on hardware. It's not a requirement from me, but would give a good starting point showing the driver worked as of this point in time.

@teburd ,
Thanks for your suggestion.
Yes, there are lots of edge cases, therefore, it took us a while to improve and verify it. Also, the dma mode(especially with chain transfer) takes more time.

Below is the test passing photo, it's with dma chain thransfer mode.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/56016699/d97d6399-19c3-4e79-b2e1-5a2849d03fb3)

kevinwang821020 commented:
Sorry to bother you reviewers, we would like to know if there are any suggestion or request for this PR.

Thanks for helping us.

ycsin commented:
Looks good to me, I think it would be better if the SPI ztest results are pasted in the OP.

cc @teburd could you take another look at this please?

tbursztyka commented on drivers/spi/Kconfig.andes_atcspi200:
```diff
@@ -11,3 +11,12 @@ config SPI_ANDES_ATCSPI200
 	depends on DT_HAS_ANDESTECH_ATCSPI200_ENABLED
 	help
 	  Enable driver for Andes ATCSPI200 SPI controller
+
```
change your commit title to: "drivers: spi: Support dma mode for atcspi200"

kevinwang821020 replied:
OK, I got it.
Thank for your recommendation, we had done the modification.

tbursztyka replied:
I still read "drivers: spi: Support dma mode"

tbursztyka commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
+	int error = 0;
+
+	data_len = data->chunk_len - 1;
+
```
you can remove that empty line as your are testing in the if below the variable you assigned previously (apply that were relevant)

kevinwang821020 replied:
Ok, thanks a lot.

tbursztyka commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -195,23 +607,57 @@ static int transceive(const struct device *dev,
 	spi_context_lock(ctx, asynchronous, cb, userdata, config);
-
 	error = configure(dev, config);
+
```
like here for instance (this change was unnecessary)

tbursztyka commented on drivers/spi/spi_andes_atcspi200.h:
```diff
@@ -11,34 +11,21 @@ LOG_MODULE_REGISTER(spi_atcspi200);
-#define REG_MCTRL       0x50
-#define REG_SLVST       0x60
-#define REG_SDCNT       0x64
-#define REG_CONFIG      0x7c
```
keel these macros

tbursztyka commented on drivers/spi/spi_andes_atcspi200.h:
```diff
@@ -11,34 +11,21 @@ LOG_MODULE_REGISTER(spi_atcspi200);
+#include <zephyr/drivers/dma.h>
+#endif
+
+#define SPI_TFMAT(base)		(base + 0x10)
```
and keep using them in these macros.

kevinwang821020 replied:
OK, no problem.
We have update our source code.
Thank you!

tbursztyka reviewed (changes requested):
Don't mix changes that are unrelated to the commit. It's better to split into proper commits then.

Commit:
drivers: spi: Support dma mode for atcspi200

1. Support the dma mode for andes_atcspi200
   and use board adp_xc7k_ae350 for testing.
2. Refine the function mechanism.

Signed-off-by: Kevin Wang <kevinwang821020@google.com>

kevinwang821020 commented:
Sorry for bothering the reviewers, we know that the version 3.5 is about to be released, and we look forward that this source code can be included in this release. Is there any good way to do this or any recommendations?

Thank you.

kevinwang821020 commented:
Due to the PR #59860(for dma driver) was merged, we rebase the source code for that.
Now with the CONFIG_DMA enable, the spi can do the transfer via dma on board adp_xc7k_ae350.

tbursztyka reviewed (commented):
You have updated your source, but not pushed force to github then?

kevinwang821020 commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
+		data->dma_tx.dma_cfg.dma_callback = NULL;
+		data->dma_tx.block_idx = 0;
+		struct dma_block_config *blk_cfg = &data->dma_tx.dma_blk_cfg;
+		const struct spi_buf *current_tx = ctx->current_tx;
```
@tbursztyka  so sorry for that, I have pushed the correct commit before, but when I do the rebase for PR #59860 last week, it seems like i do some incorrect rebase operation, so those commit was disappered.
I have corrected the git branch, it's my mistake.

Also I have update the source code for your new suggestion.

Thanks a lot!

tbursztyka commented on drivers/spi/Kconfig.andes_atcspi200:
```diff
@@ -11,3 +11,12 @@ config SPI_ANDES_ATCSPI200
+	default y
+	depends on DMA
+
+endif # SPI_ANDES_ATCSPI200
```
that should go in the second commit

kevinwang821020 replied:
OK I got it

kevinwang821020 commented:
> @kevinwang821020 any update on this one?

Hi @ycsin @teburd
Sorry for  bothering you.
I have update the source for your request before, and I just want to confirm  if my changes meet your requirements.

teburd commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -162,31 +191,412 @@ static int configure(const struct device *dev,
+static inline void dma_rx_callback(const struct device *dev, void *user_data,
+				uint32_t channel, int status)
+{
+	struct spi_atcspi200_data *data = user_data;
```
Why not use the device itself as the user_data for dma here instead if you need to get the struct device?

Can drop the self_dev pointer in data then

teburd commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -12,14 +12,39 @@
+#endif
+
 struct spi_atcspi200_data {
+	const struct device *self_dev;
```
This shouldn't be needed

kevinwang821020 replied:
OK, yes, that is more appropriate.
I have update the source code.

Thank you.

teburd reviewed (dismissed):
I'm ok with it

ycsin commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -398,43 +401,41 @@ static void spi_atcspi200_irq_handler(void *arg)
-										\
-	}
+#define SPI_INIT(n)							\
+	static struct spi_atcspi200_data spi_atcspi200_dev_data_##n = {\
```
```suggestion
	static struct spi_atcspi200_data spi_atcspi200_dev_data_##n = { \
```

kevinwang821020 replied:
Thanks, that is great!
I update the source code and rebase it.

ycsin reviewed (commented):
nit
